README.md
# time_up

Ever need to measure the elapsed wall-time for one or more bits of Ruby code,
but don't necessarily want to reach for
[Benchmark](https://ruby-doc.org/stdlib-3.0.1/libdoc/benchmark/rdoc/Benchmark.html) or roll your own ad hoc measurement code?
Try `time_up`!

This gem is especially useful for long-running processes (like test suites) that
have several time-intensive operations that are repeated over the life of the
process and that you want to measure in aggregate. (For example, to see how
much time your test suite spends creating factories, truncating the database, or
invoking a critical code path.)

Here's a [blog post about time_up](https://blog.testdouble.com/posts/2021-07-19-benchmarking-your-ruby-with-time_up/) and 
a [great example of when it can be useful](https://gist.github.com/searls/feee0b0eac7c329b390fed90c4714afb).

## Install

Just run `gem install time_up` or add time_up to your Gemfile:

```ruby
gem "time_up"
```

## Usage

Starting a timer is easy! Just call `TimeUp.start`:

```ruby
# Pass a block and time_up will only count the time while the block executes:
TimeUp.start(:factories) do
  # … create some factories
end

# Without a block, a started timer will run until you stop it:
TimeUp.start(:truncation)
# … truncate some stuff
TimeUp.stop(:truncation)
```

To get the total time that's elapsed while the timer has been running, call
`elapsed`:

```ruby
slow_time = TimeUp.elapsed(:slow_code_path)
```

Which will return a `Float` representing the number of seconds that have elapsed
while the timer is running.

Timers aggregate total elapsed time running, just like a digital stopwatch. That
means if you start a timer after it's been stopped, the additional time will be
_added_ to the previous elapsed time:

```ruby
TimeUp.start :eggs
sleep 1
puts TimeUp.stop :eggs # => ~1.0

TimeUp.start :eggs
sleep 1
# To get the elapsed time without necessarily stopping the timer, call `elapsed`
puts TimeUp.elapsed :eggs # => ~2.0

# To reset a timer to 0, call `reset` (if it's running, it'll keep running!)
TimeUp.reset :eggs
sleep 5
puts TimeUp.stop :eggs # => ~5.0
```

When passes without a block, `TimeUp.start` returns an instance of the timer,
which has its own `start`, `stop`, `elaped`, and `reset` methods. If you want to
find that instance later, you can also call `TimeUp.timer(:some_name)`. So the
above example could be rewritten as:

```ruby
egg_timer = TimeUp.start :eggs
sleep 1
puts egg_timer.stop # => ~1.0

egg_timer.start
sleep 1

# To get the elapsed time without necessarily stopping the timer, call `elapsed`
puts egg_timer.elapsed # => ~2.0

# To reset a timer to 0, call `reset` (if it's running, it'll keep running!)
egg_timer.reset
sleep 5
puts egg_timer.stop # => ~5.0
```

Finally, if you're juggling a bunch of timers, you can get a summary report of
them printed for you, like so:

```ruby
TimeUp.start :roast
sleep 0.03
TimeUp.start :veggies
sleep 0.02
TimeUp.start :pasta
sleep 0.01
TimeUp.stop_all

TimeUp.start :souffle

TimeUp.print_summary
```

Which will output something like:

```
TimeUp summary
========================
:roast   	0.07267s
:veggies 	0.03760s
:pasta   	0.01257s
:souffle*	0.00003s

* Denotes that the timer is still active
```

And if you're calling the timers multiple times and want to see some basic
statistics in the print-out, you can call `TimeUp.print_detailed_summary`, which
will produce this:

```
=============================================================================
  Name    | Elapsed | Count |   Min   |   Max   |  Mean   | Median  | 95th %
-----------------------------------------------------------------------------
:roast    | 0.08454 | 3     | 0.00128 | 0.07280 | 0.02818 | 0.01046 | 0.06657
:veggies  | 0.03779 | 1     | 0.03779 | 0.03779 | 0.03779 | 0.03779 | 0.03779
:pasta    | 0.01260 | 11    | 0.00000 | 0.01258 | 0.00115 | 0.00000 | 0.00630
:souffle* | 0.00024 | 1     | 0.00024 | 0.00025 | 0.00025 | 0.00025 | 0.00026

* Denotes that the timer is still active
```

## API

This gem defines a bunch of public methods but they're all pretty short and
straightforward, so when in doubt, I'd encourage you to [read the
code](/lib/time_up.rb).

### `TimeUp` module

`TimeUp.timer(name)` - Returns the `Timer` instance named `name` (creating it,
if it doesn't exist)

`TimeUp.start(name, [&blk])` - Starts (or restarts) a named
[Timer](#timeuptimer-class). If passed a block, will return whatever the block
evaluates to. If called without a block, it will return the timer object

`TimeUp.stop(name)` - Stops the named timer

`TimeUp.reset(name)` - Resets the named timer's elapsed time to 0, effectively
restarting it if it's currently running

`TimeUp.elapsed(name)` - Returns a `Float` of the total elapsed seconds that the
named timer has been running

`TimeUp.timings(name)` - Returns an array of each recorded start-to-stop
duration (including the current one, if the timer is running) of the named timer

`TimeUp.count(name)` - The number of times the timer has been started (including
the current timing, if the timer is running)

`TimeUp.min(name)` - The shortest recording by the timer

`TimeUp.max(name)` - The longest recording by the timer

`TimeUp.mean(name)` - The arithmetic mean of all recordings by the timer

`TimeUp.median(name)` - The median of all recordings by the timer

`TimeUp.percentile(name, percent)` - The timing for the given
[percentile](https://en.wikipedia.org/wiki/Percentile) of all recordings by the
timer

`TimeUp.total_elapsed` - Returns a `Float` of the sum of `elapsed` across all
the timers you've created (note that because you can easily run multiple logical
timers simultaneously, this figure may exceed the total time spent by the
computer)

`TimeUp.all_elapsed` - Returns a Hash of timer name keys mapped to their
`elapsed` values. Handy for grabbing a reference to a snapshot of the state of
things without requiring you to stop your timers

`TimeUp.all_stats` - Returns a Hash of timer name keys mapped to another
hash of their basic statistics (`elapsed`, `count`, `min`, `max`,
and `mean`)

`TimeUp.active_timers` - Returns an Array of all timers that are currently
running. Useful for detecting cases where you might be keeping time in multiple
places simultaneously

`TimeUp.print_summary([io])` - Pretty-prints a multi-line summary of all your
timers' total elapsed times to standard output (or the provided
[IO](https://ruby-doc.org/core-3.0.1/IO.html))

`TimeUp.print_detailed_summary([io])` - Pretty-prints a multi-line summary of
all your timers' elapsed times and basic statistics to standard output (or the
provided [IO](https://ruby-doc.org/core-3.0.1/IO.html))

`TimeUp.stop_all` - Stops all timers

`TimeUp.reset_all` - Resets all timers

`TimeUp.delete_all` - Stops and resets all timers and deletes any internal
reference to them

### `TimeUp::Timer` class

`start` - Starts the timer

`stop` - Stops the timer

`elapsed` - A `Float` of the total elapsed seconds the timer has been running

`timings` - Returns an Array of each recorded start-to-stop duration of the
timer (including the current one, if the timer is running)

`count` - The number of times the timer has been started and stopped

`min` - The shortest recording of the timer

`max` - The longest recording of the timer

`mean` - The arithmetic mean of all recorded durations of the timer

`median(name)` - The median of all recordings by the timer

`percentile(name, percent)` - The timing for the given
[percentile](https://en.wikipedia.org/wiki/Percentile) of all recordings by the
timer

`active?` - Returns `true` if the timer is running

`reset(force: false)` - Resets the timer to 0 elapsed seconds. If `force` is
true, will also stop the timer if it's running

## Code of Conduct

This project follows Test Double's [code of
conduct](https://testdouble.com/code-of-conduct) for all community interactions,
including (but not limited to) one-on-one communications, public posts/comments,
code reviews, pull requests, and GitHub issues. If violations occur, Test Double
will take any action they deem appropriate for the infraction, up to and
including blocking a user from the organization's repositories.





CHANGELOG.md
# 0.0.6

- Add `TimeUp.all_timers`. It's weird that it's not a thing.

# 0.0.5

- Add `median` and `percentile` timer statistics, and added them to
  `print_detailed_summary`

# 0.0.4

- Add `TimeUp.print_detailed_summary`

# 0.0.3

- Change the return value of TimeUp.start when passed a block to be the
  evaluated value of the block (for easier insertion into existing code without
  adding a bunch of new assignment and returns)
- Allow timer instances' `start` method to be called with a block
- Add `timings`, `count`, `min`, `max`, and `mean` methods for basic stats
  tracking
- Add `TimeUp.all_stats` to roll up all these

# 0.0.2

- Switch from a module method to Thread.current variable

# 0.0.1

- Make the gem




test/test_helper.rb
$LOAD_PATH.unshift File.expand_path("../lib", __dir__)
require "time_up"

require "minitest/autorun"


test/time_up_test.rb
require "test_helper"

class TimeUpTest < Minitest::Test
  make_my_diffs_pretty!

  def teardown
    TimeUp.delete_all
  end

  def test_that_it_has_a_version_number
    refute_nil ::TimeUp::VERSION
  end

  def test_counts_block_time
    result = TimeUp.start :alpha do
      sleep 0.1
      :some_result
    end
    assert_in_delta 0.1, TimeUp.elapsed(:alpha), 0.02
    assert_equal TimeUp.elapsed(:alpha), TimeUp.timer(:alpha).elapsed
    assert_equal :some_result, result
  end

  def test_counts_without_block
    timer = TimeUp.start :beta
    sleep 0.1
    timer.stop
    assert_in_delta 0.1, TimeUp.elapsed(:beta), 0.02
    assert_equal TimeUp.elapsed(:beta), timer.elapsed

    sleep 0.1

    timer.start
    sleep 0.1
    stop_return_value = timer.stop
    assert_in_delta 0.2, TimeUp.elapsed(:beta), 0.02
    assert_equal TimeUp.elapsed(:beta), timer.elapsed
    assert_equal TimeUp.elapsed(:beta), stop_return_value
  end

  def test_calls_start_and_stop_redundantly
    timer = TimeUp.start :charlie
    timer.start
    second_timer = TimeUp.start :charlie
    assert_same timer, second_timer
    sleep 0.1
    stop_return_1 = timer.stop
    sleep 0.1
    stop_return_2 = timer.stop
    assert_in_delta 0.1, TimeUp.elapsed(:charlie), 0.02
    assert_equal stop_return_1, stop_return_2
  end

  def test_all_timers
    t1 = TimeUp.start :t1
    t2 = TimeUp.start :t2

    assert_equal [t1, t2], TimeUp.all_timers
  end

  def test_active_timers
    t1 = TimeUp.start :t1
    t2 = TimeUp.start :t2
    t3 = TimeUp.start :t3
    t3.stop

    result = TimeUp.active_timers

    assert_equal [t1, t2], result
    assert t1.active?
    assert t2.active?
    refute t3.active?
  end

  def test_reset
    a = TimeUp.start :a
    b = TimeUp.start :b

    sleep 0.1

    assert_in_delta 0.1, TimeUp.elapsed(:a), 0.02
    assert_in_delta 0.1, TimeUp.elapsed(:b), 0.02

    a.stop
    a.reset
    TimeUp.reset(:b)
    sleep 0.1

    assert_equal 0, a.count
    assert_equal [], a.timings
    assert_equal 0.0, TimeUp.elapsed(:a)
    assert_equal TimeUp.elapsed(:a), a.elapsed
    assert_equal 1, b.count
    assert_in_delta 0.1, TimeUp.timings(:b)[0], 0.02
    assert_in_delta 0.1, TimeUp.elapsed(:b), 0.02
    assert_in_delta TimeUp.elapsed(:b), b.elapsed, 0.01
  end

  def test_stop
    TimeUp.start :z
    result = TimeUp.stop :z

    sleep 0.05

    assert_in_delta 0.0, TimeUp.elapsed(:z), 0.02
    assert_equal TimeUp.elapsed(:z), result
  end

  def test_raises_when_named_timer_doesnt_exist
    e = assert_raises(TimeUp::Error) { TimeUp.elapsed(:fake) }
    assert_equal "No timer named :fake", e.message

    e = assert_raises(TimeUp::Error) { TimeUp.stop(:blah) }
    assert_equal "No timer named :blah", e.message

    e = assert_raises(TimeUp::Error) { TimeUp.reset(:kek) }
    assert_equal "No timer named :kek", e.message

    e = assert_raises(TimeUp::Error) { TimeUp.timings(:hrm) }
    assert_equal "No timer named :hrm", e.message
  end

  def test_raises_if_name_isnt_a_string_or_symbol
    e = assert_raises(TimeUp::Error) { TimeUp.start(42) }
    assert_equal "Timer name must be a String or Symbol", e.message
  end

  def test_elapsed_aggregates_past_plus_active_unless_reset
    TimeUp.start :apple do
      sleep 0.1
    end

    TimeUp.stop :apple

    assert_in_delta 0.1, TimeUp.elapsed(:apple), 0.02

    TimeUp.start :apple
    sleep 0.1

    assert_in_delta 0.2, TimeUp.elapsed(:apple), 0.02
  end

  def test_calling_reset_on_a_running_timer_keeps_running
    TimeUp.start :a
    sleep 0.1
    TimeUp.reset :a
    sleep 0.2
    assert_in_delta 0.2, TimeUp.elapsed(:a), 0.02
  end

  def test_total_elapsed
    TimeUp.start :foo
    TimeUp.start :bar
    TimeUp.start :baz

    sleep 0.1

    assert_in_delta 0.3, TimeUp.total_elapsed, 0.03

    TimeUp.stop :bar
    sleep 0.1
    TimeUp.stop_all

    assert_in_delta 0.5, TimeUp.total_elapsed, 0.04

    sleep 0.1

    assert_in_delta 0.5, TimeUp.total_elapsed, 0.06
    assert_in_delta 0.2, TimeUp.timer(:foo).elapsed, 0.06
    assert_in_delta 0.1, TimeUp.timer(:bar).elapsed, 0.06
    assert_in_delta 0.2, TimeUp.timer(:baz).elapsed, 0.06
  end

  def test_total_elapsed_with_no_timers
    assert_equal 0.0, TimeUp.total_elapsed
  end

  def test_delete_all
    t = TimeUp.start(:t)

    TimeUp.delete_all

    refute_same t, TimeUp.timer(:t)
    assert_equal 0.0, t.elapsed
    refute t.active?
  end

  def test_timer_name_is_accessible
    t = TimeUp.start(:jerry)

    assert_equal :jerry, t.name
  end

  def test_string_timers_dont_blow_up
    t = TimeUp.start("cool neat")

    sleep 0.1

    assert_equal "cool neat", t.name
    assert_in_delta 0.1, TimeUp.elapsed("cool neat"), 0.02
  end

  def test_all_elapsed
    TimeUp.start :t1
    sleep 0.1
    t2 = TimeUp.start :t2
    sleep 0.1
    t2.stop
    TimeUp.start :t3
    sleep 0.1

    result = TimeUp.all_elapsed

    assert_equal 3, result.size
    assert_in_delta 0.3, result[:t1], 0.02
    assert_in_delta 0.1, result[:t2], 0.02
    assert_in_delta 0.1, result[:t3], 0.02
  end

  def test_print_summary
    TimeUp.start :roast
    sleep 0.03
    TimeUp.start :veggies
    sleep 0.02
    TimeUp.start :pasta
    sleep 0.01
    TimeUp.stop_all
    TimeUp.start :souffle

    string_io = StringIO.new
    TimeUp.print_summary(string_io)

    lines = string_io.tap(&:rewind).read.split("\n")

    assert_equal "", lines[0]
    assert_equal "TimeUp summary", lines[1]
    assert_equal "========================", lines[2]
    assert_match(/:roast   	0\.0[567]\d{3}s/, lines[3])
    assert_match(/:veggies 	0\.0[234]\d{3}s/, lines[4])
    assert_match(/:pasta   	0\.0[123]\d{3}s/, lines[5])
    assert_match(/:souffle\*	0\.00\d{3}s/, lines[6])
    assert_equal "", lines[7]
    assert_equal "* Denotes that the timer is still active", lines[8]
  end

  def test_print_detailed_summary
    TimeUp.start :roast
    sleep 0.03
    TimeUp.start :veggies
    sleep 0.02
    TimeUp.start :pasta
    sleep 0.01
    TimeUp.stop_all
    TimeUp.start(:roast) { sleep 0.01 }
    TimeUp.start(:roast) { sleep 0.001 }
    10.times { TimeUp.start(:pasta) {} }
    TimeUp.start :souffle

    string_io = StringIO.new
    TimeUp.print_detailed_summary(string_io)

    lines = string_io.tap(&:rewind).read.split("\n")

    [
      "",
      "=============================================================================",
      "  Name    | Elapsed | Count |   Min   |   Max   |  Mean   | Median  | 95th % ",
      "-----------------------------------------------------------------------------",
      /^:roast    \| 0.0[789]\d{3} \| 3     \| 0.00\d{3} \| 0.0[678]\d{3} \| 0.0[123]\d{3} \| 0.0[123]\d{3} \| 0.0[567]\d{3}/,
      /^:veggies  \| 0.0[234]\d{3} \| 1     \| 0.0[234]\d{3} \| 0.0[234]\d{3} \| 0.0[234]\d{3} \| 0.0[234]\d{3} \| 0.0[234]\d{3}/,
      /^:pasta    \| 0.0[012]\d{3} \| 11    \| 0.00000 \| 0.0[012]\d{3} \| 0.00\d{3} \| 0.00\d{3} \| 0.00\d{3}/,
      /^:souffle\* \| 0.00\d{3} \| 1     \| 0.00\d{3} \| 0.00\d{3} \| 0.00\d{3} \| 0.00\d{3} \| 0.00\d{3}/,
      "",
      "* Denotes that the timer is still active"
    ].each.with_index do |expected, i|
      assert_match expected, lines[i]
    end
  end

  def test_timings
    TimeUp.start(:z) { sleep 0.1 }

    assert_equal 1, TimeUp.timings(:z).size
    assert_in_delta 0.1, TimeUp.timings(:z)[0], 0.01
    assert_equal TimeUp.timer(:z).timings, TimeUp.timings(:z)

    TimeUp.start(:z)
    sleep 0.05

    ongoing_timings = TimeUp.timings(:z)
    assert_equal 2, ongoing_timings.size
    assert_in_delta 0.1, ongoing_timings[0], 0.01
    assert_in_delta 0.05, ongoing_timings[1], 0.01
  end

  def test_empty_stats
    timer = TimeUp.timer(:a)

    assert_equal 0, timer.elapsed
    assert_equal TimeUp.elapsed(:a), timer.elapsed
    assert_equal 0, timer.count
    assert_equal TimeUp.count(:a), timer.count
    assert_equal 0, timer.timings.size
    assert_equal TimeUp.timings(:a).size, timer.timings.size
    assert_nil timer.min
    assert_nil TimeUp.min(:a)
    assert_nil timer.max
    assert_nil TimeUp.max(:a)
    assert_nil timer.mean
    assert_nil TimeUp.mean(:a)
    assert_nil timer.median
    assert_nil TimeUp.median(:a)
    assert_nil timer.percentile(95)
    assert_nil TimeUp.percentile(:a, 95)
  end

  def test_basic_stats_tracking
    timer = TimeUp.timer(:a)
    timer.start { sleep 0.1 }
    timer.start { sleep 0.2 }
    sleep 0.1
    timer.start { sleep 0.1 }
    timer.start { sleep 0.05 }

    assert_in_delta 0.45, timer.elapsed, 0.03
    assert_equal 4, timer.count
    assert_equal 4, timer.timings.size
    assert_in_delta 0.1, timer.timings[0], 0.01
    assert_in_delta 0.2, timer.timings[1], 0.01
    assert_in_delta 0.1, timer.timings[2], 0.01
    assert_in_delta 0.05, timer.timings[3], 0.01
    assert_in_delta 0.05, timer.min, 0.01
    assert_in_delta 0.2, timer.max, 0.01
    assert_in_delta 0.1125, timer.mean, 0.03
    assert_in_delta 0.1, timer.median, 0.02
    assert_equal 0, timer.percentile(0)
    assert_equal 0, timer.percentile(-1.0)
    assert_in_delta 0.2, timer.percentile(100.0), 0.01
    assert_in_delta 0.2, timer.percentile(110.0), 0.01
    assert_in_delta 0.2, timer.percentile(95), 0.03
    assert_in_delta 0.095, timer.percentile(30), 0.03
  end

  def test_all_stats
    TimeUp.start(:a) { sleep 0.05 }
    TimeUp.start(:b) { sleep 0.05 }
    TimeUp.start(:a) { sleep 0.1 }

    result = TimeUp.all_stats

    a = TimeUp.timer(:a)
    b = TimeUp.timer(:b)

    assert_equal({
      a: {
        elapsed: a.elapsed,
        count: 2,
        min: a.min,
        max: a.max,
        mean: a.mean,
        median: a.median,
        "95th": a.percentile(95)
      },
      b: {
        elapsed: b.elapsed,
        count: 1,
        min: b.min,
        max: b.max,
        mean: b.mean,
        median: b.median,
        "95th": b.percentile(95)
      }
    }, result)
  end
end


lib/time_up.rb
require_relative "time_up/version"

module TimeUp
  class Error < StandardError; end

  Thread.current[:time_up_timers] = {}

  def self.timer(name)
    __timers[name] ||= Timer.new(name)
  end

  # Delegate methods
  def self.start(name, &blk)
    timer(name).start(&blk)
  end

  [
    :stop,
    :reset,
    :elapsed,
    :timings,
    :count,
    :min,
    :max,
    :mean,
    :median
  ].each do |method_name|
    define_singleton_method method_name do |name|
      __ensure_timer(name)
      __timers[name].send(method_name)
    end
  end

  def self.percentile(name, percentage)
    __ensure_timer(name)
    __timers[name].percentile(percentage)
  end

  # Interrogative methods
  def self.total_elapsed
    __timers.values.sum(&:elapsed)
  end

  def self.all_elapsed
    __timers.values.map { |timer|
      [timer.name, timer.elapsed]
    }.to_h
  end

  def self.all_timers
    __timers.values
  end

  def self.all_stats
    __timers.values.map { |timer|
      [timer.name, {
        elapsed: timer.elapsed,
        count: timer.count,
        min: timer.min,
        max: timer.max,
        mean: timer.mean,
        median: timer.median,
        "95th": timer.percentile(95)
      }]
    }.to_h
  end

  def self.active_timers
    __timers.values.select(&:active?)
  end

  def self.print_summary(io = $stdout)
    longest_name_length = __timers.values.map { |t| t.name.inspect.size }.max
    summaries = __timers.values.map { |timer|
      name = "#{timer.name.inspect}#{"*" if timer.active?}".ljust(longest_name_length + 1)
      "#{name}\t#{"%.5f" % timer.elapsed}s"
    }
    io.puts <<~SUMMARY

      TimeUp summary
      ========================
      #{summaries.join("\n")}

      #{"* Denotes that the timer is still active\n" if __timers.values.any?(&:active?)}
    SUMMARY
  end

  def self.print_detailed_summary(io = $stdout)
    cols = {
      names: ["Name"],
      elapsed: ["Elapsed"],
      count: ["Count"],
      min: ["Min"],
      max: ["Max"],
      mean: ["Mean"],
      median: ["Median"],
      "95th": ["95th %"]
    }
    __timers.values.each { |timer|
      cols[:names] << "#{timer.name.inspect}#{"*" if timer.active?}"
      cols[:elapsed] << "%.5f" % timer.elapsed
      cols[:count] << timer.count.to_s
      cols[:min] << "%.5f" % timer.min
      cols[:max] << "%.5f" % timer.max
      cols[:mean] << "%.5f" % timer.mean
      cols[:median] << "%.5f" % timer.median
      cols[:"95th"] << "%.5f" % timer.percentile(95)
    }

    widths = cols.map { |name, vals|
      [name, vals.map(&:length).max]
    }.to_h

    rows = cols[:names].size.times.map { |i|
      if i == 0
        cols.keys.map { |name|
          cols[name][i].center(widths[name])
        }
      else
        cols.keys.map { |name|
          cols[name][i].ljust(widths[name])
        }
      end
    }

    full_width = widths.values.sum + (rows[0].size - 1) * 3
    io.puts <<~SUMMARY

      #{"=" * full_width}
      #{rows[0].join(" | ")}
      #{"-" * full_width}
      #{rows[1..-1].map { |row| row.join(" | ") }.join("\n")}

      #{"* Denotes that the timer is still active\n" if __timers.values.any?(&:active?)}
    SUMMARY
  end

  # Iterative methods
  def self.stop_all
    __timers.values.each(&:stop)
  end

  def self.reset_all
    __timers.values.each(&:reset)
  end

  def self.delete_all
    __timers.values.each { |t| t.reset(force: true) }
    Thread.current[:time_up_timers] = {}
  end

  # Internal methods
  def self.__timers
    Thread.current[:time_up_timers] ||= {}
  end

  def self.__ensure_timer(name)
    raise Error.new("No timer named #{name.inspect}") unless __timers[name]
  end

  class Timer
    attr_reader :name

    def initialize(name)
      validate!(name)
      @name = name
      @start_time = nil
      @total_elapsed = 0.0
      @past_timings = []
    end

    def start(&blk)
      @start_time ||= now
      if blk
        blk.call.tap do
          stop
        end
      else
        self
      end
    end

    def stop
      if @start_time
        duration = now - @start_time
        @past_timings.push(duration)
        @total_elapsed += duration

        @start_time = nil
      end
      @total_elapsed
    end

    def elapsed
      if active?
        @total_elapsed + (now - @start_time)
      else
        @total_elapsed
      end
    end

    def reset(force: false)
      if force
        @start_time = nil
      elsif !@start_time.nil?
        @start_time = now
      end
      @total_elapsed = 0.0
      @past_timings = []
    end

    def count
      timings.size
    end

    def min
      timings.min
    end

    def max
      timings.max
    end

    def mean
      times = timings
      return if times.empty?
      times.sum / times.size
    end

    def median
      times = timings.sort
      return if times.empty?
      (times[(times.size - 1) / 2] + times[times.size / 2]) / 2.0
    end

    def percentile(percent)
      times = timings.sort
      return if times.empty?
      return 0 if percent <= 0
      return max if percent >= 100
      return times.first if times.size == 1
      position = (percent / 100.0) * (times.size - 1)

      partial_ratio = position - position.floor
      whole, partial = times[position.floor, 2]
      whole + (partial - whole) * partial_ratio
    end

    def timings
      if active?
        @past_timings + [now - @start_time]
      else
        @past_timings
      end
    end

    def active?
      !!@start_time
    end

    private

    def validate!(name)
      unless name.is_a?(Symbol) || name.is_a?(String)
        raise Error.new("Timer name must be a String or Symbol")
      end
    end

    def now
      Process.clock_gettime(Process::CLOCK_MONOTONIC)
    end
  end
end
